crates/core/tedge_mapper/src/mapper.rs

```diff
@@ -85,14 +89,121 @@ impl Mapper {
+                            .topic
+                            .name
+                            .strip_prefix("tedge-internal/alarms/")
+                            .expect("Expected tedge-internal/alarms prefix")
+                            .to_string();
+                        old_alarms_map
+                            .insert(alarm_id, message);
+                    }
+                }
+            }
+        }
+
+        // Compare the differences between alarms in tedge/alarms topic to the ones in tedge-internal/alarms topic
+        for (key, old_message) in old_alarms_map.into_iter() {
+            match pending_alarms_map.get(&key) {
+                // If an alarm that is present in tedge-internal/alarms topic is not present in tedge/alarms topic,
+                // it is assumed to have been cleared while the mapper process was down
+                None => {
+                    let topic = Topic::new_unchecked(format!("tedge/alarms/{}", key).as_str());
+                    let message = Message::new(&topic, vec![]).with_retain();
+                    // Recreate the clear alarm message and add it to the pending alarms list to be processed later
+                    pending_alarms_map.insert(key, message);
+                }
+
+                // If the payload of a message received from tedge/alarms is same as one received from tedge-internal/alarms,
+                // it is assumed to be one that was already processed earlier and hence removed from the pending alarms list.
+                Some(message) => {
+                    if message.payload_bytes() == old_message.payload_bytes() {
+                        pending_alarms_map.remove(&key);
+                    }
+                }
+            }
+        }
+
+        // Once all the pending alarms are identified, process them
+        for (_key, message) in pending_alarms_map.into_iter() {
+            self.process_message(message).await;
+        }
+    }
+
+    async fn process_message(&mut self, message: Message) {
+        let converted_messages = self.converter.convert(&message);
+        for converted_message in converted_messages.into_iter() {
+            let _ = self.output.send(converted_message).await;
+
+            // Persist a copy of alarm messages to an internal topic for reconciliation on next restart
+            if message.topic.name.starts_with("tedge/alarms") {
+                let alarm_id = message
+                    .topic
+                    .name
+                    .strip_prefix("tedge/alarms/")
+                    .expect("Expected tedge/alarms prefix")
+                    .to_string();
+
+                let topic =
+                    Topic::new_unchecked(format!("tedge-internal/alarms/{}", alarm_id).as_str());
+                let alarm_copy =
+                    Message::new(&topic, message.payload_bytes().to_owned()).with_retain();
+                let _ = self.output.send(alarm_copy).await;
+            }
+        }
```

Comment:
This needs to be moved inside the converter.

When an alarm is received, the converter must return *two* messages:
* the translated alarm,
* and the internal copy.